indexing-service/src/main/java/org/apache/druid/indexing/compact/CompactionSupervisor.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.compact;
+
+import org.apache.druid.indexing.overlord.DataSourceMetadata;
+import org.apache.druid.indexing.overlord.supervisor.Supervisor;
+import org.apache.druid.indexing.overlord.supervisor.SupervisorReport;
+import org.apache.druid.indexing.overlord.supervisor.SupervisorStateManager;
+import org.apache.druid.indexing.overlord.supervisor.autoscaler.LagStats;
+import org.apache.druid.java.util.common.DateTimes;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.server.coordinator.AutoCompactionSnapshot;
+
+/**
+ * Supervisor for compaction of a single datasource.
+ */
+public class CompactionSupervisor implements Supervisor
+{
+  private static final Logger log = new Logger(CompactionSupervisor.class);
+
+  private final String dataSource;
+  private final CompactionScheduler scheduler;
+  private final CompactionSupervisorSpec supervisorSpec;
+
+  public CompactionSupervisor(
+      CompactionSupervisorSpec supervisorSpec,
+      CompactionScheduler scheduler
+  )
+  {
+    this.supervisorSpec = supervisorSpec;
+    this.scheduler = scheduler;
+    this.dataSource = supervisorSpec.getSpec().getDataSource();
+  }
+
+  @Override
+  public void start()
+  {
+    if (supervisorSpec.isSuspended()) {
+      log.info("Suspending compaction for dataSource[%s].", dataSource);
+      scheduler.stopCompaction(dataSource);
```

Comment:
When is this case encountered, i.e. the supervisor for the datasource is starting but the compaction for the same datasource needs to be stopped?

Reply:
This is because of the way the supervisor framework works. This code is triggered when `start()` is called for a suspended supervisor.